### PR TITLE
Fix interface fetching with upgrades (#19858)

### PR DESCRIPTION
**Pull request by samuel-williams-da**
Manual forwardport of #19858 and relevant part of #19676

### CODE REVIEW

File: sdk/daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/Preprocessor.scala [hunk 1]
```diff
@@ -219,11 +220,25 @@ private[engine] final class Preprocessor(
...
-  ): Result[speedy.Command] =
-    safelyRun(pullPackage(List(cmd.templateId))) {
+      cmd: ReplayCommand
+  ): Result[speedy.Command] = {
+    def templateAndInterfaceIds =
+      cmd match {
+        case ReplayCommand.Create(templateId, _) => List(templateId)
+        // Exercise includes interface package in main-3.x, to fix a bug with retroactive instances
+        // These aren't supported in 3x. Unclear if Exeercise should include the id.
+        // TODO: Clarify
```
Comment (samuel-williams-da) [inline sdk/daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/Preprocessor.scala:230]: Note here, as https://github.com/digital-asset/daml/pull/19822  wasn't forward ported
Note also that it implements a test that we can no longer crib from. Once we have a decision on this PR, we should port over the test.

File: sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltinFun.scala [hunk 2]
```diff
@@ -1266,13 +1269,27 @@ private[lf] object SBuiltinFun {
...
       if (tplId == tyCon) {
-        SOptional(Some(record))
+        Control.Value(SOptional(Some(record)))
+      } else if (tplId.qualifiedName == tyCon.qualifiedName) {
+        val (tplIdPkgName, _) = machine.tmplId2PackageNameVersion(tplId)
+        val (tyConPkgName, _) = machine.tmplId2PackageNameVersion(tyCon)
+        if (tplIdPkgName == tyConPkgName) {
+          importValue(machine, tplId, record.toUnnormalizedValue) { templateArg =>
+            Control.Value(SOptional(Some(templateArg)))
+          }
```
Comment (samuel-williams-da) [inline sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltinFun.scala:1287]: This isn't good enough for 3x, but its a larger task to fix. For now, I'm just aligning the branches

File: sdk/canton/community/base/src/main/scala/com/digitalasset/canton/data/ActionDescription.scala [hunk 3]
```diff
@@ -381,6 +384,7 @@ object ActionDescription extends HasProtocolVersionedCompanion[ActionDescription
     override def pretty: Pretty[ExerciseActionDescription] = prettyOfClass(
       param("input contract id", _.inputContractId),
       param("template id", _.templateId),
+      param("interface id", _.interfaceId),
```
Comment (rgugliel-da) [inline sdk/canton/community/base/src/main/scala/com/digitalasset/canton/data/ActionDescription.scala:387]: paramIfDefined ?

Comment (samuel-williams-da) [reply sdk/canton/community/base/src/main/scala/com/digitalasset/canton/data/ActionDescription.scala:387]: Ah, I haven't done this on 2x, as there were other optional fields that also just showed "None" explicitly when missing
I can update it here, but it'll be inconsistent

Comment (rgugliel-da) [reply sdk/canton/community/base/src/main/scala/com/digitalasset/canton/data/ActionDescription.scala:387]: If relevant, I'd change for paramIfDefined: inconsistency is not an issue here

File: sdk/daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessor.scala [hunk 4]
```diff
@@ -260,10 +260,16 @@ private[lf] final class CommandPreprocessor(
...
           ) =>
         unsafePreprocessExerciseByKey(templateId, contractKey, choiceId, argument, strict = true)
-      case command.ReplayCommand.Fetch(tmplId, coid) =>
+      case command.ReplayCommand.Fetch(tmplId, ifaceIdOpt, coid) =>
         val cid = valueTranslator.unsafeTranslateCid(coid)
-        discard(handleLookup(pkgInterface.lookupTemplate(tmplId)))
-        speedy.Command.FetchTemplate(tmplId, cid)
+        ifaceIdOpt match {
+          case Some(ifaceId) =>
+            val _ = handleLookup(pkgInterface.lookupInterfaceInstance(ifaceId, tmplId))
```
Comment (paulbrauner-da) [inline sdk/daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/CommandPreprocessor.scala:267]: Nit: use `discard` here or `val _ =` below for consistency.

**Review by matthiasS-da — APPROVED** [review]
The Canton parts look good to me.

File: sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltinFun.scala (hunk 2 above)
Comment (paulbrauner-da) [reply sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltinFun.scala:1287]: Maybe leave a note about why this is not good enough.

File: sdk/daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/Preprocessor.scala (hunk 1 above)
Comment (paulbrauner-da) [reply sdk/daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/Preprocessor.scala:230]: I'm not sure I fully understand this comment:
>Note here, as #19822 wasn't forward ported

It seems to me this is (partly at least) a port of #19822

> Note also that it implements a test that we can no longer crib from. Once we have a decision on this PR, we should port over the test.

What kind of decision should we reach?

**Pull request merged**